discovery/kubernetes/client_metrics.go

```diff
@@ -0,0 +1,225 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kubernetes
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/metrics"
+	"k8s.io/client-go/util/workqueue"
+)
+
+const (
+	cacheMetricsNamespace     = metricsNamesapce + "_cache"
+	workqueueMetricsNamespace = metricsNamesapce + "_workqueue"
+)
+
+var (
+	// Metrics for client-go adapters observi client's HTTP requests
+	clientGoRequestResultMetric = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_result_total",
+			Help:      "Request results of the client-go on the Kubernetes API.",
+		},
+		[]string{"status_code", "method"},
+	)
+	clientGoRequestLatencyMetric = prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace: metricsNamesapce,
+			Name:      "http_request_duration_seconds",
+			Help:      "Request latency of the client-go on Kubernetes API.",
+		},
+		[]string{"verb", "endpoint"},
+	)
+
+	// Definition of metrics for client-go cache metrics provider
+	clientGoCacheListTotalMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "cache_list_total",
+			Help:      "Total number of client-go Kubernets API calls.",
+		},
+	)
+	clientGoCacheListDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_duration_seconds",
+			Help:      "Duration of a Kubernetes API call in seconds.",
+		},
+	)
+	clientGoCacheItemsInListCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "list_items_count",
+			Help:      "Count of items in a list from the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watches_total",
+			Help:      "Count of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheShortWatchesCountMetric = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "short_watches_total",
+			Help:      "Count of short watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheWatchesDurationMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_duration_seconds",
+			Help:      "Duration of watches on the Kubernetes API.",
+		},
+	)
+	clientGoCacheItemsInWatchesCountMetric = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Namespace: cacheMetricsNamespace,
+			Name:      "watch_items",
```

Comment:
IIUC this metric observes the number of events received in the watch loops. Maybe `watch_events` is more appropriate?